internal/generate/crd/crd.go

```diff
@@ -0,0 +1,332 @@
+	resource scaffold.Resource
+}
+
+// NewCRDGo returns a CRD generator configured to generate CustomResourceDefintion
+// manifests from Go API files.
+func NewCRDGo(cfg genutil.Config) genutil.Generator {
+	g := crdGenerator{
+		Config:       cfg,
+		isOperatorGo: true,
+	}
+	if g.InputDir == "" {
+		g.InputDir = scaffold.ApisDir
+	}
+	if g.OutputDir == "" {
+		g.OutputDir = scaffold.CRDsDir
+	}
+	return g
+}
+
+// NewCRDGo returns a CRD generator configured to generate a
+// CustomResourceDefintion manifest from scratch using data in resource.
+func NewCRDNonGo(cfg genutil.Config, resource scaffold.Resource) genutil.Generator {
+	g := crdGenerator{
+		Config:       cfg,
+		resource:     resource,
+		isOperatorGo: false,
+	}
+	if g.InputDir == "" {
+		g.InputDir = scaffold.CRDsDir
+	}
+	if g.OutputDir == "" {
+		g.OutputDir = scaffold.CRDsDir
+	}
+	return g
+}
+
+func (g crdGenerator) validate() error {
+	if g.InputDir == "" {
+		return errors.New("input dir cannot be empty")
+	}
+	if g.InputDir == "" {
+		return errors.New("output dir cannot be empty")
+	}
+	if !g.isOperatorGo {
+		if err := g.resource.Validate(); err != nil {
+			return errors.Wrap(err, "resource is invalid:")
+		}
+	}
+	return nil
+}
+
+// Generate generates CRD manifests and writes them to g.OutputDir.
+func (g crdGenerator) Generate() (err error) {
+	if err = g.validate(); err != nil {
+		return errors.Wrap(err, "validation error")
+	}
+	var fileMap map[string][]byte
+	if g.isOperatorGo {
+		fileMap, err = g.generateGo()
+	} else {
+		fileMap, err = g.generateNonGo()
+	}
+	if err != nil {
+		return errors.Wrap(err, "error generating CRD manifests")
+	}
+	if err = os.MkdirAll(g.OutputDir, fileutil.DefaultDirFileMode); err != nil {
+		return errors.Wrapf(err, "error mkdir %s", g.OutputDir)
+	}
+	for fileName, b := range fileMap {
+		path := filepath.Join(g.OutputDir, fileName)
+		if err := ioutil.WriteFile(path, b, fileutil.DefaultFileMode); err != nil {
+			return errors.Wrap(err, "error writing CRD manifests")
+		}
+	}
+	return nil
+}
+
+func getFileNameForResource(r scaffold.Resource) string {
+	return fmt.Sprintf("%s_%s_crd.yaml", r.FullGroup, r.Resource)
+}
+
+// generateNonGo generates CRDs for Go projects using Go API files.
+func (g crdGenerator) generateGo() (map[string][]byte, error) {
+	fileMap := map[string][]byte{}
+	// Generate files in the generator's cache so we can modify the file name
+	// and annotations.
+	defName := "output:crd:cache"
+	cacheOutputDir := string(filepath.Separator) + filepath.Clean(g.OutputDir)
+	rawOpts := []string{
+		"crd",
+		fmt.Sprintf("paths=%s/...", fileutil.DotPath(g.InputDir)),
+		fmt.Sprintf("%s:dir=%s", defName, cacheOutputDir),
+	}
+	cachedGen := genutil.NewCachedGenerator()
+	cachedGen.AddOutputRule(defName, genutil.OutputToCachedDirectory{})
+	if err := cachedGen.Run(rawOpts); err != nil {
+		return nil, err
+	}
+	cache := genutil.GetCache()
+	infos, err := afero.ReadDir(cache, cacheOutputDir)
```

Comment:
Is it possible that `controller-gen` will create a hierarchy directory structure in the future? If so, should we do the afero equivalent of `filepath.Walk()`?

Reply:
I don't think so since we control the custom output rule `OutputToCachedDirectory`, to which you must explicitly provide the dir to output files.

Reply:
Right, but say `cacheOutputDir` is `/tmp/myOutputDir`. Is it guaranteed that files will always be  written in _that_ directory, or is it possible that at some point in the future, controller-gen could create files in subdirectories (e.g. `/tmp/myOutputDir/<group>/<resource>.yaml`)

It's probably not anything we really need to worry about now since a change like that would be caught in our unit test. But something I thought about nonetheless.

Reply:
It is guaranteed unless we change our code. Any change in path would come from modifying an [`OutputRule`](https://github.com/kubernetes-sigs/controller-tools/blob/master/pkg/genall/output.go#L77-L78) and the only `OutputRule` our generator is configured with is one we've written, which creates a file at the top level of the supplied dir.